### DISCUSSION TITLE
Obtaining VisualStyleGroup from a Style

### DISCUSSION THREAD

**Discussion by HowardPWeiss** [unoplatform/uno#17322] [Q&A]
Linux, .Net 8.0,  UNO 5+

I use a straightforward routine in C# to obtain style groups for a control

```
public static object VisualStateGroupsForObject(object obj)
{
	Microsoft.UI.Xaml.FrameworkElement fe = obj as Microsoft.UI.Xaml.FrameworkElement;
	IList<VisualStateGroup> vsg = VisualStateManager.GetVisualStateGroups(fe);
	return vsg;
}

If, in Xaml, I define a button as follows

<Button Name="btnRead" Command="{Binding btnRead_Click}" IsEnabled="{Binding Path=bReadEnabled}" 
            Height="27" Width="95" TabIndex="12" Margin="107,20,0,0" HorizontalAlignment="Left" VerticalAlignment="Top" 
            Content ="Read" >
                <VisualStateManager.VisualStateGroups> 
                    <VisualStateGroup x:Name="CommonStates">
                 
                        <VisualState x:Name="Normal" />

                        <VisualState x:Name="PointerOver">
                            <VisualState.Setters>
                                <Setter Target="ContentPresenter.Foreground" Value="Yellow" />  
                            </VisualState.Setters>
                        </VisualState>

                        <VisualState x:Name="Pressed" />

                        <VisualState x:Name="Disabled"/>
                        
                        <VisualState x:Name="Focused"/>

                        <VisualState x:Name="PointerFocused"/>

                    </VisualStateGroup>
                </VisualStateManager.VisualStateGroups>
        </Button>```

and call my routine to obtain the visualStateGroups,  I see the vsg is [Ilist]: {Microsoft.UI.Xaml.VisualStateGroup[1]}

However, if I define a style and specify that the button uses that style

<Style TargetType ="Button" x:Key="TestStyle0">
        <Setter Property="Background" Value="Red" />
        <Setter Property="Foreground" Value="Blue" />
        <Setter Property="Template">
            <Setter.Value>
                <ControlTemplate TargetType="Button">
                    <Border x:Name="RootElement">
                        <VisualStateManager.VisualStateGroups> 
                            <VisualStateGroup x:Name="CommonStates">
                            
                                <VisualState x:Name="Normal" />
    
                                <VisualState x:Name="PointerOver">
                                    <VisualState.Setters>
                                        <Setter Target="ContentPresenter.Foreground" Value="Yellow" />  
                                    </VisualState.Setters>
                                </VisualState>
    
                                <VisualState x:Name="Pressed" />
    
                                <VisualState x:Name="Disabled"/>
                                
                                <VisualState x:Name="Focused"/>
    
                                <VisualState x:Name="PointerFocused"/>
    
                            </VisualStateGroup>
                        </VisualStateManager.VisualStateGroups>
    
                        <ContentPresenter x:Name="ContentPresenter"
                            Margin="{TemplateBinding Padding}"
                            Content="{TemplateBinding Content}"
                            ContentTemplate="{TemplateBinding ContentTemplate}"
                            ContentTransitions="{TemplateBinding ContentTransitions}"
                            FontFamily="{TemplateBinding FontFamily}"
                            FontSize="{TemplateBinding FontSize}"
                            FontWeight="{TemplateBinding FontWeight}"
                            Foreground="{TemplateBinding Foreground}"
                            Background="{TemplateBinding Background}"
                            CharacterSpacing="{TemplateBinding CharacterSpacing}"
                            HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
                            VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}" />
                    </Border>	
                </ControlTemplate>
            </Setter.Value>
        </Setter>
    </Style>``


       <Button Name="btnSolve" Command="{Binding btnSolve_Click}" IsEnabled="{Binding Path=bSolveEnabled}" Height="27" Width="76" TabIndex="8" Margin="210,20,0,0" HorizontalAlignment="Left" VerticalAlignment="Top" Style = "{StaticResource TestStyle0}">Solve</Button>

```

I find that vsg is  [Ilist]: {Microsoft.UI.Xaml.VisualStateGroup[0]}.  That is, no VisualState groups are returned even though the Style defines visit state groups

How should I obtain the VisualStateGroups for the styled button?

**Comment by jeromelaban** [comment]
In Uno, for performance reasons, visual state groups are not materialized eagerly for performance reasons. `FindName` may be able to find them, but otherwise, there's no way to force the materialization. 

Why do you need to access the states this way?

**Reply by HowardPWeiss** [reply]
I am working on the behavior of the buttons in my application.  The VisualStateManager information controls the button appearance as I move the mouse over the buttons, etc.  In UNO, when the VisualStateManager is in a ControlTemplate, the returned VisualStateGroup list is an IList with zero entries. I am forced to read the .xaml file to understand the behavior

By the way, in a WPF application on VisualStudio2022, I find the .baml file in the Resources section of the .dll file.  For a UNO application,  I cannot find the .baml file in the .dll on either Windows or Linux.  There is a class BamlRead2006 in System.Windows.BAML2006 which may or may not work to read the .BAML file from the resources.

Also, the Live Visual Tree tool in VIsual Studio 2022 allows me to navigate the Visual Tree in WPF application (including displayint the XAML for a ControlTemplate).  Do you know of an equivalent tool (or Visual Studio Code extension) for UNO applications running on Linux

![XAMLTest2 ilSpy showing mainwindow bam](https://github.com/unoplatform/uno/assets/91093577/26c12fd6-ce35-4602-a1af-a6e66540e83d)

**Reply by jeromelaban** [reply]
Uno does not generate any baml files that you can use to replicate this behavior.

> What does "visual state groups are not materialized eagerly for performance reasons' mean? Not eagerly == lazy? Is there some way that I could invoke the lazy access. I added a new class derived from Button and declared my button which uses a style to be of that class. I then overrode the "On Apply Template" method. I had hoped that this method would be invoked after the visual state groups were materialized not eagerly. Unfortunately, GetVisualStateGroups still returns an empty IList

It's not a common use case to read the visual state groups contents through code, and Uno [does not materialize the content of those](https://github.com/unoplatform/uno/blob/5f7bba38a8fd64f5b23e0088e41c0cf9cb2df4b5/src/Uno.UI/UI/Xaml/VisualState.cs#L165) to make control creation faster. `VisualStateGroups` are always created, though, so you should be able to get them. \

In all cases, make sure to create a version of your code that works with WinAppSDK, then test it with Uno and see if the behavior differs, the post of a fully working repro here so we can properly track it.

**Comment by HowardPWeiss** [comment]
Jerome

Please clarify

What does "visual state groups are not materialized eagerly for performance reasons' mean? Not eagerly == lazy? Is there some way that I could invoke the lazy access.  I added a new class derived from Button and declared my button which uses a style to be of that class.  I then overrode the "On Apply Template" method.  I had hoped that this method would be invoked after the visual state groups were materialized not eagerly.  Unfortunately, GetVisualStateGroups still returns an empty IList

How would I potentially use FindName? I tried naming the ControlTemplate and using FindName(<control tempalte name>) but FindName returned null

You never indicated whether the .baml file can be found in the .dll for my UnoProject

**Comment by HowardPWeiss** [comment]
Jerome

Here is code to read the VisualStateGroup for a Button with a Style containing a VisualStateGroup. myVisual is the button object obtaing using FindName

```
	// Enumerate all the descendants of the visual object.
        public static void EnumVisual(object myVisual)
        {
			DependencyObject myVisualDO = myVisual as DependencyObject;
			int childCount = VisualTreeHelper.GetChildrenCount(myVisualDO);
			for (int i = 0; i < VisualTreeHelper.GetChildrenCount(myVisualDO); i++)
            {
				// Retrieve child visual at specified index value.
				object childVisual = VisualTreeHelper.GetChild(myVisualDO, i);
				FrameworkElement cvf = childVisual as FrameworkElement;
				//object list = VisualStateManager.GetVisualStateGroups(cvf);
				IList<VisualStateGroup> list = VisualStateManager.GetVisualStateGroups(cvf);
				//IList<Microsoft.UI.Xaml.VisualStateGroup> list2 = VisualStateManager.GetVisualStateGroups(cvf);
				Debug.WriteLine(String.Format("childVisual {0} EnumVisual List has {1} elements", childVisual, list.Count)); 
				if (i < list.Count)
				{
					VisualStateGroup vsg = list[i];
					Debug.WriteLine(string.Format(" vsg Name {0}", vsg.Name));
					Debug.WriteLine(string.Format("     Number Of States {0}", vsg.States.Count));
					for (int jj = 0; jj < vsg.States.Count; jj++)
					{
						Debug.WriteLine(string.Format("     State[0] Name {1}", jj, vsg.States[jj].Name));
					}
					// Enumerate children of the child visual object.
					EnumVisual(childVisual);
				}
			}
        }
```

This resolved my problem
